catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/DuplicateCommands.java

```diff
@@ -138,20 +152,37 @@ protected void duplicateInBatches(CatalogFacade queryFacade, CatalogFacade inges
                     rejectedExecutionHandler);
             console.printf("Running a maximum of %d threads during replication.%n", multithreaded);
 
-            printProgressAndFlush(start, totalWanted, ingestedCount.get());
+            printProgressAndFlush(start,
+                    Math.max(totalWanted, initialMetacards.size()),
+                    ingestedCount.get());
             int index;
-            while ((index = queryIndex.addAndGet(batchSize)) <= totalWanted) {
-                final int i = index;
+            while (!done.get()) {
+                index = queryIndex.addAndGet(batchSize);
+                final int taskIndex = index;
 
                 executorService.submit(() -> {
-                    final SourceResponse response = query(queryFacade,
-                            filter,
-                            i,
-                            getQuerySizeFromIndex(totalWanted, i));
-                    if (response != null) {
-                        ingestMetacards(ingestFacade, getMetacardsFromSourceResponse(response));
+                    int querySize = (int) getQuerySizeFromIndex(totalWanted, taskIndex);
+                    if (querySize < 1) {
+                        // If we don't need any more metacards, we're finished
+                        done.set(true);
+                        return;
                     }
-                    printProgressAndFlush(start, totalWanted, ingestedCount.get());
+                    List<Metacard> metacards = ResultIterable.resultIterable(queryFacade::query,
+                            queryTemplate.apply(taskIndex),
+                            querySize)
+                            .stream()
+                            .map(Result::getMetacard)
+                            .collect(Collectors.toList());
+
+                    if (metacards.size() < querySize) {
```

Comment:
I thought it was necessary to query until you received 0 results because of filtering.

Reply:
@clockard Yes and no. Thanks to this comment I have discovered a whole new whale of an issue 🎉  -- @coyotesqrl is tracking now. 

Filtering basically completely ruins the few guarantees we could rely on before (like that result size was indicative of if you were done).